### PR TITLE
Busbud Coding Challenge Submission

### PR DESCRIPTION
**Pull request by Clorr**
### Submission for Clorr to the Busbud Coding Challenge

Demo at https://tranquil-chamber-64444.herokuapp.com/suggestions?q=Londo&latitude=43.70011&longitude=-79.4163

# Limitations
Query parameters order is fixed to `?q=&lat=&lon=` . I did a _regexp_ parsing for demo, but this can be easily changed.

# Improvments to be considered
Number of suggestions returned could be limited
The load of the data can be improved, but as it is done once at startup, I did not dig into the possibilities.

# Tests
Added some and refactored the existing

# Denormalization
The TSV file is parsed into a JSON file for index. This adds an additional step at each source file change. 

I prefer this solution to the direct parsing of TSV because providing a raw & untested file to a service may block the starting of this service if the input has some errors

The denormalization is done by running `app-denormalize.js`

# Streams
As I don't know streams yet, I'm not sure of the overhead and of the use of map/mapSync. So I sticked to safest uses.

# Score
Score is poorly implemented, by a lack of idea on how to tackle this part. (Levenshtein being useless on a prefix as it returns only the substring length)

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -1,16 +1,52 @@
...
-    }));
-  } else {
-    res.end();
-  }
-}).listen(port, '127.0.0.1');
-
-console.log('Server running at http://127.0.0.1:%d/suggestions', port);
\ No newline at end of file
+	try {
+		var urlChunks = req.url.split("&");
```
Comment (nchaulet) [inline app.js]: use the module from node `querystring` ?

Comment (Clorr) [reply app.js]: Possibly `querystring` is a better choice, this is the part where my bad knowledge of node libraries comes in.
I did not want to import `url` as I thought it was overkill for this, but I was not aware of `querystring`

**Comment by Clorr** [comment]
I've done some cleanups, and hopefully, improvments.

Functionally, there is now a scoring related to alternative names when the match is not made on ascii name. When this is the case, I prefix the full name with altName => fullName.
Ex: "xpn => Brampton, ON, Canada"

@nchaulet , URL parsing is now done with a Regex, I hope you don't mind. Your feedback made me think that you have high expectations on the code quality. 
Please note that this is a challenge submission and this code is not intended to be put in prod. This is more a demo of:
- How I tried to satisfy to most of the functional requirements
- How I tackled the general problem
- My programming skills, which are certainly lower for Node JS than other people that use it daily

------------------------------------------------------------------------

FYI, as a added a denormalisation step, here is the output log:

Loaded 7237 lines
... => (names collisions put below)
map has 145468 entries sharded into 58 groups of 16185 max size
data has 7237 entries
done!

------------------------------------------------------------------------

Name collision on Langley, BC, Canada
Name collision on Langley, BC, Canada
Name collision on Bella Vista, AR, USA
Name collision on Bella Vista, AR, USA
Name collision on Citrus Park, FL, USA
Name collision on Lakewood Park, FL, USA
Name collision on Lakewood Park, FL, USA
Name collision on Country Club Estates, GA, USA
Name collision on Country Club Estates, GA, USA
Name collision on Fairfield, IL, USA
Name collision on Lowell, IN, USA
Name collision on Mount Vernon, IN, USA
Name collision on Monticello, KY, USA
Name collision on Monticello, KY, USA
Name collision on Morehead, KY, USA
Name collision on Morehead, KY, USA
Name collision on Oak Grove, KY, USA
Name collision on Oak Grove, KY, USA
Name collision on Oakdale, LA, USA
Name collision on Oakdale, LA, USA
Name collision on Riverside, MD, USA
Name collision on Riverside, MD, USA
Name collision on Robinwood, MD, USA
Name collision on Robinwood, MD, USA
Name collision on Woodlawn, MD, USA
Name collision on Woodlawn, MD, USA
Name collision on Holly Springs, MS, USA
Name collision on Holly Springs, MS, USA
Name collision on Lake Norman of Catawba, NC, USA
Name collision on Summerfield, NC, USA
Name collision on Summerfield, NC, USA
Name collision on Northfield, NJ, USA
Name collision on Geneva, OH, USA
Name collision on Northridge, OH, USA
Name collision on Northridge, OH, USA
Name collision on Red Lion, PA, USA
Name collision on Shiloh, PA, USA
Name collision on Boiling Springs, SC, USA
Name collision on Boiling Springs, SC, USA
Name collision on Centerville, SC, USA
Name collision on Centerville, SC, USA
Name collision on Oak Grove, SC, USA
Name collision on Red Hill, SC, USA
Name collision on Red Hill, SC, USA
Name collision on Saint Andrews, SC, USA
Name collision on Saint Andrews, SC, USA
Name collision on Mount Carmel, TN, USA
Name collision on Mount Carmel, TN, USA
Name collision on Center, TX, USA
Name collision on Center, TX, USA
Name collision on Cleveland, TX, USA
Name collision on Cleveland, TX, USA
Name collision on Centerville, IA, USA
Name collision on Centerville, IA, USA
Name collision on Fairfield, IL, USA
Name collision on Lynwood, IL, USA
Name collision on Lynwood, IL, USA
Name collision on Lowell, IN, USA
Name collision on Mount Vernon, IN, USA
Name collision on Bloomingdale, NJ, USA
Name collision on Bloomingdale, NJ, USA
Name collision on Northfield, NJ, USA
Name collision on Prospect Park, NJ, USA
Name collision on Prospect Park, NJ, USA
Name collision on Centerport, NY, USA
Name collision on Centerport, NY, USA
Name collision on Inwood, NY, USA
Name collision on Inwood, NY, USA
Name collision on Lakeview, NY, USA
Name collision on Lakeview, NY, USA
Name collision on Manorville, NY, USA
Name collision on Manorville, NY, USA
Name collision on Oakdale, NY, USA
Name collision on Oakdale, NY, USA
Name collision on Tonawanda, NY, USA
Name collision on Woodbury, NY, USA
Name collision on Woodbury, NY, USA
Name collision on Geneva, OH, USA
Name collision on Northwood, OH, USA
Name collision on Northwood, OH, USA
Name collision on Doylestown, PA, USA
Name collision on Doylestown, PA, USA
Name collision on Maple Glen, PA, USA
Name collision on Maple Glen, PA, USA
Name collision on Middletown, PA, USA
Name collision on Middletown, PA, USA
Name collision on Mountain Top, PA, USA
Name collision on Mountain Top, PA, USA
Name collision on Red Lion, PA, USA
Name collision on Shiloh, PA, USA
Name collision on Wilson, PA, USA
Name collision on Wilson, PA, USA
Name collision on Alta Sierra, CA, USA
Name collision on Alta Sierra, CA, USA
Name collision on Bay Point, CA, USA
Name collision on Bay Point, CA, USA
Name collision on El Cerrito, CA, USA
Name collision on El Cerrito, CA, USA
Name collision on Las Flores, CA, USA
Name collision on Las Flores, CA, USA
Name collision on Live Oak, CA, USA
Name collision on Live Oak, CA, USA
Name collision on Lee Acres, NM, USA
Name collision on Lee Acres, NM, USA
Name collision on Fairwood, WA, USA
Name collision on Fairwood, WA, USA
Name collision on Hawaiian Paradise Park, HI, USA
Name collision on Hawaiian Paradise Park, HI, USA
Name collision on Citrus Park, FL, USA
Name collision on Tonawanda, NY, USA
Name collision on Lake Norman of Catawba, NC, USA
Name collision on Oak Grove, SC, USA

**Comment by Clorr** [comment]
I realised while writing to Michael that it was maybe better to use the real cities file, so I pushed it.

-----
Here is the new log of denormalizer:

Loaded 7495 lines => after filtering
... => see name collisions
map has 368686 entries sharded into 67 groups of 36931 max size
data has 7495 entries
done!

---------

**Langley, BC, Canada** => **Langley (Pop: 23606), BC, Canada**
**Langley, BC, Canada** => **Langley (Pop: 93726), BC, Canada**
**Sainte-Catherine, QC, Canada** => **Sainte-Catherine (Pop: 16762), QC, Canada**
**Sainte-Catherine, QC, Canada** => **Sainte-Catherine (Pop: 16211), QC, Canada**
**Lakewood Park, FL, USA** => **Lakewood Park (Pop: 11323), FL, USA**
**Lakewood Park, FL, USA** => **Lakewood Park (Pop: 11323), FL, USA**
**Country Club Estates, GA, USA** => **Country Club Estates (Pop: 7555), GA, USA**
**Country Club Estates, GA, USA** => **Country Club Estates (Pop: 8545), GA, USA**
**Fairfield, IL, USA** => **Fairfield (Pop: 5040), IL, USA**
**Lowell, IN, USA** => **Lowell (Pop: 7914), IN, USA**
**Mount Vernon, IN, USA** => **Mount Vernon (Pop: 6550), IN, USA**
**Monticello, KY, USA** => **Monticello (Pop: 6090), KY, USA**
**Monticello, KY, USA** => **Monticello (Pop: 6103), KY, USA**
**Morehead, KY, USA** => **Morehead (Pop: 7045), KY, USA**
**Morehead, KY, USA** => **Morehead (Pop: 7622), KY, USA**
**Oak Grove, KY, USA** => **Oak Grove (Pop: 7444), KY, USA**
**Oak Grove, KY, USA** => **Oak Grove (Pop: 7989), KY, USA**
**Riverside, MD, USA** => **Riverside (Pop: 6425), MD, USA**
**Riverside, MD, USA** => **Riverside (Pop: 6507), MD, USA**
**Robinwood, MD, USA** => **Robinwood (Pop: 5024), MD, USA**
**Robinwood, MD, USA** => **Robinwood (Pop: 6918), MD, USA**
**Woodlawn, MD, USA** => **Woodlawn (Pop: 37879), MD, USA**
**Woodlawn, MD, USA** => **Woodlawn (Pop: 6334), MD, USA**
**Holly Springs, MS, USA** => **Holly Springs (Pop: 7901), MS, USA**
**Holly Springs, MS, USA** => **Holly Springs (Pop: 7610), MS, USA**
**Lake Norman of Catawba, NC, USA** => **Lake Norman of Catawba (Pop: 5075), NC, USA**
**Northfield, NJ, USA** => **Northfield (Pop: 8521), NJ, USA**
**Geneva, OH, USA** => **Geneva (Pop: 6447), OH, USA**
**Northridge, OH, USA** => **Northridge (Pop: 7572), OH, USA**
**Northridge, OH, USA** => **Northridge (Pop: 8487), OH, USA**
**Red Lion, PA, USA** => **Red Lion (Pop: 6321), PA, USA**
**Shiloh, PA, USA** => **Shiloh (Pop: 11218), PA, USA**
**Boiling Springs, SC, USA** => **Boiling Springs (Pop: 5076), SC, USA**
**Boiling Springs, SC, USA** => **Boiling Springs (Pop: 8219), SC, USA**
**Centerville, SC, USA** => **Centerville (Pop: 6586), SC, USA**
**Centerville, SC, USA** => **Centerville (Pop: 5288), SC, USA**
**Oak Grove, SC, USA** => **Oak Grove (Pop: 8183), SC, USA**
**Red Hill, SC, USA** => **Red Hill (Pop: 13223), SC, USA**
**Red Hill, SC, USA** => **Red Hill (Pop: 13223), SC, USA**
**Mount Carmel, TN, USA** => **Mount Carmel (Pop: 5298), TN, USA**
**Mount Carmel, TN, USA** => **Mount Carmel (Pop: 5425), TN, USA**
**Center, TX, USA** => **Center (Pop: 5727), TX, USA**
**Center, TX, USA** => **Center (Pop: 5211), TX, USA**
**Cleveland, TX, USA** => **Cleveland (Pop: 7936), TX, USA**
**Cleveland, TX, USA** => **Cleveland (Pop: 7858), TX, USA**
**Centerville, IA, USA** => **Centerville (Pop: 5372), IA, USA**
**Centerville, IA, USA** => **Centerville (Pop: 5745), IA, USA**
**Fairfield, IL, USA** => **Fairfield (Pop: 5255), IL, USA**
**Lynwood, IL, USA** => **Lynwood (Pop: 9280), IL, USA**
**Lynwood, IL, USA** => **Lynwood (Pop: 7713), IL, USA**
**Lowell, IN, USA** => **Lowell (Pop: 9450), IN, USA**
**Mount Vernon, IN, USA** => **Mount Vernon (Pop: 7208), IN, USA**
**Bloomingdale, NJ, USA** => **Bloomingdale (Pop: 8215), NJ, USA**
**Bloomingdale, NJ, USA** => **Bloomingdale (Pop: 7714), NJ, USA**
**Northfield, NJ, USA** => **Northfield (Pop: 8070), NJ, USA**
**Prospect Park, NJ, USA** => **Prospect Park (Pop: 5794), NJ, USA**
**Prospect Park, NJ, USA** => **Prospect Park (Pop: 5953), NJ, USA**
**Centerport, NY, USA** => **Centerport (Pop: 5457), NY, USA**
**Centerport, NY, USA** => **Centerport (Pop: 5508), NY, USA**
**Inwood, NY, USA** => **Inwood (Pop: 9792), NY, USA**
**Inwood, NY, USA** => **Inwood (Pop: 10082), NY, USA**
**Lakeview, NY, USA** => **Lakeview (Pop: 5625), NY, USA**
**Lakeview, NY, USA** => **Lakeview (Pop: 5615), NY, USA**
**Manorville, NY, USA** => **Manorville (Pop: 14314), NY, USA**
**Manorville, NY, USA** => **Manorville (Pop: 13729), NY, USA**
**Oakdale, NY, USA** => **Oakdale (Pop: 8107), NY, USA**
**Oakdale, NY, USA** => **Oakdale (Pop: 7974), NY, USA**
**Woodbury, NY, USA** => **Woodbury (Pop: 8907), NY, USA**
**Woodbury, NY, USA** => **Woodbury (Pop: 10879), NY, USA**
**Geneva, OH, USA** => **Geneva (Pop: 6051), OH, USA**
**Northwood, OH, USA** => **Northwood (Pop: 5469), OH, USA**
**Northwood, OH, USA** => **Northwood (Pop: 5350), OH, USA**
**Doylestown, PA, USA** => **Doylestown (Pop: 8301), PA, USA**
**Doylestown, PA, USA** => **Doylestown (Pop: 8146), PA, USA**
**Maple Glen, PA, USA** => **Maple Glen (Pop: 7635), PA, USA**
**Maple Glen, PA, USA** => **Maple Glen (Pop: 6742), PA, USA**
**Middletown, PA, USA** => **Middletown (Pop: 9117), PA, USA**
**Middletown, PA, USA** => **Middletown (Pop: 7441), PA, USA**
**Red Lion, PA, USA** => **Red Lion (Pop: 6055), PA, USA**
**Shiloh, PA, USA** => **Shiloh (Pop: 11217), PA, USA**
**Wilson, PA, USA** => **Wilson (Pop: 7564), PA, USA**
**Wilson, PA, USA** => **Wilson (Pop: 7781), PA, USA**
**Alta Sierra, CA, USA** => **Alta Sierra (Pop: 7047), CA, USA**
**Alta Sierra, CA, USA** => **Alta Sierra (Pop: 6911), CA, USA**
**El Cerrito, CA, USA** => **El Cerrito (Pop: 23549), CA, USA**
**El Cerrito, CA, USA** => **El Cerrito (Pop: 5100), CA, USA**
**Las Flores, CA, USA** => **Las Flores (Pop: 6037), CA, USA**
**Las Flores, CA, USA** => **Las Flores (Pop: 5971), CA, USA**
**Live Oak, CA, USA** => **Live Oak (Pop: 17158), CA, USA**
**Live Oak, CA, USA** => **Live Oak (Pop: 8531), CA, USA**
**Vincent, CA, USA** => **Vincent (Pop: 15922), CA, USA**
**Lee Acres, NM, USA** => **Lee Acres (Pop: 5858), NM, USA**
**Lee Acres, NM, USA** => **Lee Acres (Pop: 5858), NM, USA**
**Fairwood, WA, USA** => **Fairwood (Pop: 19102), WA, USA**
**Fairwood, WA, USA** => **Fairwood (Pop: 7905), WA, USA**
**Klahanie, WA, USA** => **Klahanie (Pop: 10674), WA, USA**
**Lake Norman of Catawba, NC, USA** => **Lake Norman of Catawba (Pop: 7411), NC, USA**
**Oak Grove, SC, USA** => **Oak Grove (Pop: 10291), SC, USA**
**Vincent, CA, USA** => **Vincent (Pop: 15922), CA, USA**
**Klahanie, WA, USA** => **Klahanie (Pop: 10674), WA, USA**

**Comment by Clorr** [comment]
### Released
## Asked
- /suggestions endpoint (partial/complete/lat-lon)
- Functional tests
- Sort with score _=> Note that score is poorly implemented. (And a Levensthein is useless since we don't do approximate match. A Levenshtein on a prefix is just the size of the substring)_
- Deployed on Heroku
- All in JS
- Pull request

## Not explicitly asked
- Error codes
- Unicode handling

### Performance
- The code is not done to get the maximum performance, but it has a design in that direction:
  - Search is done through a denormalized object where each prefix is precomputed (since the quantity of data is low, this a viable solution), the speed then relies on the JS internal `object["str"]` performance, which is hopefully fast since this one the most used feature in JS. (Actually, I'm not aware of the real internal behavior, but I suspect it to be fast...). 
  - The data has been sharded by `prefix.length`, this is mostly to search in smaller objects rather than in a big object containing everything, but the impact on performance has not been tested.

- Other part that may have an impact have not been particularly optimized (sorting, serialization, url parsing)

- Scoring is poorly done so, at least, it has been made to be fast. Therefore a score is computed on a single object. If we had to score objects relatively to each other, it would have had a potential impact

- The program has been split into a denormalization step, and a server module because parsing the source file is costly, and should not be done at each server startup. The speed of this part is then less important. The load of the file can be done in a better manner, but that is part of the improvments.

### Possible improvments
- Score
- Load of JSON data file
- GZip _=> can be provided with a simple pipe to zlib_
- Remote fetch of source in order to get fresh data
- Approximate search
- Performance tuning
- Improve disambuiguation

### Found "Easter eggs"

# README.md
- Different ports _=> was not intentional ?_
- SomeRandomCityInTheMiddleOfNowhere _=> if we look for a city, we should find it even if it is in the middle of nowhere ;-)_
- Misspelling "in Geonames provides..."

# app.js
- Writes 404 unconditionally
- indexOf === 0 is fine when there is a match, but parses all the string if no match
- json response without proper header
- listen on 127.0.0.1 does not work on Heroku _=> this one was a bit sadistic :-p_
- console.log written before the server has started listening

# .nvmrc
- ... _=> had trouble with nvm, I'm on windows_

# package.json
- old versions
- dependencies not separated from devDependencies (I found this while reading another pull request)
- npm test does not work because of the 'node' command  missing (I found this while reading another pull request)
_=> Note: I still use 6.12, because I had trouble with 8.9.1 on Windows. I had a similar error than this one: https://github.com/remy/nodemon/issues/1124 ._

# test/suggestions.js
- regex interverted with tested string
- SomeRandomCityInTheMiddleOfNowhere also here
- unprotected JSON.Parse heavily crashes the tests if the response is not JSON  _=>(at least my VS code does not like it...)_

# data/README.md & data/cities_canada-usa.tsv
- alternate names are quite a mess with unicode names...
- Almost similar cities...
_=> Also I realize that the real cities file is as messy as the one from the challenge :-p_

**Comment by Clorr** [comment]
I played a bit with `eslint`, in order to give my code a better Node JS formalism. It is still a bit raw, but I hope it is better so.

**Pull request closed without merging**